### PR TITLE
Add rules sampler to the tracer 

### PR DESCRIPTION
**Pull request by cgilmour**
This required quite a lot of changes, to be able to build, pass data around to places that now need it, and has limited tests.
Opening PR to get the functionality reviewed, and decide what remaining things will be done in this PR vs a followup one.

### CODE REVIEW

File: src/sample.cpp [hunk 1]
```diff
@@ -89,5 +89,44 @@ std::shared_ptr<SampleProvider> sampleProviderFromOptions(const TracerOptions& o
   return std::shared_ptr<SampleProvider>{new DiscardRateSampler(1.0 - options.sample_rate)};
 }
 
+RulesSampler::RulesSampler() : sampling_limiter_(getRealTime, 100, 100.0, 1) {}
+
+void RulesSampler::addRule(RuleFunc f) { sampling_rules_.push_back(f); }
+
+SampleResult RulesSampler::sample(const std::string& environment, const std::string& service,
+                                  const std::string& name, uint64_t trace_id) {
+  auto limit_result = sampling_limiter_.allow();
```
Comment (brettlangdon) [inline src/sample.cpp]: we should be matching on the rule first, we only need to rate limit traces which would be sampled (p1).

File: src/span_buffer.cpp [hunk 2]
```diff
@@ -12,6 +12,8 @@ const std::string sampling_priority_metric = "_sampling_priority_v1";
 const std::string datadog_origin_tag = "_dd.origin";
 const std::string datadog_hostname_tag = "_dd.hostname";
 const std::string event_sample_rate_metric = "_dd1.sr.eausr";
+const std::string rules_sampler_applied_rate = "_dd.rule_psr";
+const std::string rules_sampler_limiter_rate = "_dd.limit_psr";
```
Comment (brettlangdon) [inline src/span_buffer.cpp]: don't forget about `_dd.agent_psr`, for when we use priority sampling to sample

File: src/sample.cpp [hunk 3]
```diff
@@ -26,68 +23,89 @@ uint64_t maxIdFromKeepRate(double rate) {
...
+
+void RulesSampler::addRule(RuleFunc f) { sampling_rules_.push_back(f); }
+
+SampleResult RulesSampler::sample(const std::string& environment, const std::string& service,
+                                  const std::string& name, uint64_t trace_id) {
+  auto rule_result = match(service, name);
+  if (!rule_result.matched) {
+    return priority_sampler_.sample(environment, service, trace_id);
   }
+  auto limit_result = sampling_limiter_.allow();
```
[the remark this answers is not in the document]
Comment (cgilmour) [reply src/sample.cpp]: Updated to behave as pseudo-code describes.

File: src/sample.cpp [hunk 4]
```diff
@@ -26,68 +23,89 @@ uint64_t maxIdFromKeepRate(double rate) {
...
+  if (!limit_result.allowed) {
+    return priority_sampler_.sample(environment, service, trace_id);
```
[the remark this answers is not in the document]
Comment (cgilmour) [reply src/sample.cpp]: Updated and added the tests for this.

**Pull request merged**